### PR TITLE
Use an embedded browser window to obtain an authorization code

### PR DESCRIPTION
**Pull request by fifthdegree**
This allows us to actually capture the authorization code instead of having to manually paste it in.

### CODE REVIEW

**Review by akallabeth — COMMENTED** [review]
not really sure to pull in qt as a dependency is a good idea.

maybe something like https://github.com/babelouest/iddawc/blob/v1.1.8/examples/microsoft_example.c is sufficient?

**Comment by akallabeth** [comment]
not entirely sure we want a whole qt as dependency, but it might be ok for a client to (optionally) have this. (e.g. add it only to `SDL` client as that is already `C++` without changing the defaults for `FreeRDP` library itself)

File: client/common/webview.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+			if (pair.size() != 2 || pair[0] != "code")
+				continue;
+
+			QByteArray code = pair[1].toUtf8();
+			*codeptr = (char*)calloc(1, code.size() + 1);
+			strcpy(*codeptr, code.constData());
+			break;
+		}
+		qApp->exit();
```
Comment (hardening) [inline client/common/webview.cpp:54]: Perhaps we could pass the QApplication as parameter in the ctor instead.

**Comment by hardening** [comment]
I'd bet that creating the QApplication in FreeRDP will bother any program that uses FreeRDP and Qt...

**Comment by fifthdegree** [comment]
> I'd bet that creating the QApplication in FreeRDP will bother any program that uses FreeRDP and Qt...

We can just check if `qApp` is already initialized. But if you guys decide that you want to scrap this idea, I do have another (not as good) idea. Turns out the client id used by client.wvd.microsoft.com has "https://login.microsoftonline.com/common/oauth2/nativeclient" as one of its 90 valid redirect URIs. That way, at least the url shows up in the browser bar, so it's easier to copy/paste

**Comment by akallabeth** [comment]
@fifthdegree I´ve been looking into this a bit and found https://github.com/webview/webview which is a wrapper around platform specific webview stuff.
might be exactly what we need here (open a webview, get result code) and does have c bindings (so you don´t need a c++ compiler for the `FreeRDP` core project (the SDL client and the proxy modules are separate `CMake` projects you don´t have to build)
the previous mentioned library seems to be only able to handle the parsing of the oauth-2 parts and does not contain a webview

**Comment by fifthdegree** [comment]
The issue with that wrapper is that it doesn't wrap the functionality needed to intercept requests in any way

**Comment by fifthdegree** [comment]
@akallabeth, @hardening I've moved things around a bit.

**Review by akallabeth — APPROVED** [review]
LGTM

**Comment by akallabeth** [comment]
@freerdp-bot please test this nice thing!

File: client/SDL/CMakeLists.txt [hunk 2]
```diff
@@ -98,10 +98,29 @@ set(LIBS
         freerdp-client
     )
 
+option(WITH_WEBVIEW "Build with QtWebEngine support for AAD login popup browser" OFF)
+if (WITH_WEBVIEW)
+	include(FindPkgConfig)
+	pkg_check_modules(QT Qt5WebEngineWidgets)
```
Comment (akallabeth) [inline client/SDL/CMakeLists.txt:104]: might be a good idea to just add `REQUIRED` here and remove the found check below. If we already request `WITH_WEBVIEW=ON` manually then expect it is available and if not fail.

**Comment by akallabeth** [comment]
looks good (except a `clang-format` missing).
with a follow up pr we can split out `sdl_webview_get_aad_auth_code` to a separate library and wrap up different implementations (gtk, qt, windows, apple, ...)

File: client/SDL/CMakeLists.txt [hunk 3]
```diff
@@ -98,10 +98,25 @@ set(LIBS
...
+  include_directories(${QT_INCLUDE_DIRS})
+  list(APPEND LIBS ${QT_LIBRARIES})
+endif()
+
 add_executable(${PROJECT_NAME}
 	${SRCS}
     )
 
+if (QT_FOUND)
+  target_compile_options(${PROJECT_NAME} PRIVATE -fPIC)
```
Comment (akallabeth) [inline client/SDL/CMakeLists.txt:117]: that is already handled with `CMAKE_POSITION_INDEPENDENT_CODE` (or the corresponding target option)

**Review by akallabeth — APPROVED** [review]

**Review by hardening — APPROVED** [review]

File: client/SDL/CMakeLists.txt (hunk 3 above)
Comment (fifthdegree) [reply client/SDL/CMakeLists.txt:117]: Apparently, cmake will only enable the -fPIE flag which is not good enough for qt

Comment (akallabeth) [reply client/SDL/CMakeLists.txt:117]: `-v`? 
* I can compile perfectly without this line (and start `sdl-freerdp`)
* this option will only work with `gcc` and `clang` but fail with ms compilers
* https://stackoverflow.com/questions/15978608/qt-compile-your-code-with-fpic-or-fpie

**Comment by akallabeth** [comment]
@freerdp-bot test

File: client/SDL/CMakeLists.txt (hunk 3 above)
Comment (akallabeth) [reply client/SDL/CMakeLists.txt]: @fifthdegree but I´ve found the issue, using `find_package` with the `Qt::WebEngineWidgets` target properly sets all required options. When using `pkg-config` you have to manually set all of these, so I´ve pushed a fix here.

**Pull request merged**